---
title: "AMAZE Analysis - Monica N Mweetwa"
---
Set working directory
```{r, setup}
knitr::opts_knit$set(root.dir = 'C:/Users/Monica/Dropbox/AMAZE/')
#setwd("C:/Users/Monica/Dropbox/AMAZE")
#setwd("/Users/monica_mweetwa/Library/CloudStorage/Dropbox/AMAZE")
```

Load packages
```{r include=FALSE}
library("data.table")
library("phyloseq")
library("ggbeeswarm")
library("ggrepel")
library("vegan")
library("tidyverse")
library(seqinr)
library(ape)
library(msa) 
library(stringr)
library(upstartr)
library(microbiome)
library(readxl)
```

Load data
```{r include=FALSE}
count_data <- read_csv("Inputs/Duod_Aspirate/AMAZE_duodenum_bacterial_counts.csv")
library(readr)
sample_metadata <- read_csv("Inputs/Duod_Aspirate/metdat_duod.csv")
taxa_data <- read.csv("Inputs/Duod_Aspirate/AMAZE_duodenum_taxonomy.csv")
```

Create phyloseq object
```{r}
##change order so that bar plots shoe pre-intervention before post-interventions
levels(as.factor(sample_metadata$Timepoint))
sample_metadata$Timepoint <- factor(sample_metadata$Timepoint, levels = c("Pre-Intervention", "Post-Intervention", ""))
levels(as.factor(sample_metadata$Timepoint))

#filter for duodenal samples only
sample_metadata <- sample_metadata %>%
  filter(!is.na(SampleID))

otu_mat <- count_data %>%
  tibble::column_to_rownames("SampleID") 

tax_mat <- taxa_data %>% 
  tibble::column_to_rownames("SampleID")

sample_id <- sample_metadata %>% 
  tibble::column_to_rownames("SampleID") 

#Transform otu and tax tables into matrix (sample table can be left as data frame)
otu_mat <- as.matrix(round(otu_mat))
tax_mat <- as.matrix(tax_mat)
OTU = otu_table(otu_mat , taxa_are_rows = TRUE)
TAX = tax_table(tax_mat)
samples = sample_data(sample_id)

amaze_duod <- phyloseq(otu_table(OTU), tax_table(TAX),sample_data(samples))
amaze_duod
```

Add sequence information to phyloseq object using the taxonomy tables
```{r}
RefTable <- taxa_data %>%
  select(SampleID, ASVSeq) %>%
  rename(SampleID = "seq.name",
         ASVSeq = "seq.text")
library(phylotools)
dat2fasta(RefTable, outfile = "Inputs/Duod_Aspirate/AMAZE_duodenum_sequences.fasta")
```

Create phylogenetic tree
```{r}
sequence <- readAAStringSet("Inputs/Duod_Aspirate/AMAZE_duodenum_sequences.fasta")   
amaze_duod <- phyloseq(otu_table(OTU), tax_table(TAX),sample_data(samples), refseq(sequence))

## Perform multiple sequence alignment
my_alignment <- msa(sequence)

## Compute distance matrix
my_alignment_sequence <- msaConvert(my_alignment, type="seqinr::alignment")

distance_alignment <- dist.alignment(my_alignment_sequence)

## compute phylogenetic tree using neighbor joining
Tree <- bionj(distance_alignment)
```

Add tree to phyloseq object
```{r}
#Rename tip.labels of tree to match the names of phyloseq taxa
#Tree$tip.label <- str_sub(Tree$tip.label, end=-1)
Tree2 <- Tree
write.tree(Tree2, "Inputs/Duod_Aspirate/amaze_tree_duodenum.tre")
phy_tree(amaze_duod) <- Tree2
amaze_duod
```

Clean phyloseq object
1. Remove wrongly assigned taxa
```{r}
ps0 <- subset_samples(amaze_duod, InterventionArm != "") #Remove samples that are not paired 
ps0 = subset_taxa(amaze_duod, Kingdom == "Bacteria") # remove those with no taxonomic assignment
ps0 = subset_taxa(ps0, Family != "Mitochondria") #remove those assigned as Mitochondria Family
ps0 = subset_taxa(ps0, Class != "Chloroplast") #remove those assigned as Chloroplast Class 
ps0_duodenum <- ps0
ps0_duodenum
```

```{r}
save(ps0_duodenum, file = "Inputs/RData/AMAZE_duodenum.RData")
```

Agglomerate at Genus level
```{r}
ps0.genus.duod <- tax_glom(ps0_duodenum, taxrank = "Genus", NArm = TRUE)
#Rename taxa with genus names
genus.names <- make.names(tax_table(ps0.genus.duod)[,'Genus'], unique = TRUE)
taxa_names(ps0.genus.duod) <- genus.names
ps0.genus.duod 
```

Save RData file
```{r}
save(ps0.genus.duod, file = "Inputs/RData/AMAZE_duodenum_genus.RData")
```
