---
title: "AMAZE - Merging Stool and Duodenal Microbiome Data"
---

Set working directory
```{r setup}
    knitr::opts_knit$set(root.dir = normalizePath("C:/Users/Monica/Dropbox/AMAZE")) 
    #knitr::opts_knit$set(root.dir = normalizePath("~/Library/CloudStorage/Dropbox/AMAZE")) 
```

Load packages
```{r include=FALSE}
library("data.table")
library("phyloseq")
library("tidyverse")
library(stringr)
library(export)
library(upstartr)
library(msa)
```

Load Datasets
```{r}
load("Inputs/RData/AMAZE_fecal_genus.RData")
load("Inputs/RData/AMAZE_duodenum_genus.RData")
```

Create new phyloseq objects without phylogenetic trees because of different tip lengths
```{r}
#1. Duodenum
tax  = tax_table(ps0.genus.duod)
otu  = otu_table(ps0.genus.duod)
sam  = sample_data(ps0.genus.duod)
seq  = refseq(ps0.genus.duod)
amaze.duod.genus2 = phyloseq(otu, tax, sam, seq)
amaze.duod.genus2

#2. Fecal
tax  = tax_table(ps0.genus.fecal)
otu  = otu_table(ps0.genus.fecal)
sam  = sample_data(ps0.genus.fecal)
seq  = refseq(ps0.genus.fecal)
amaze.fecal.genus2 = phyloseq(otu, tax, sam, seq)
amaze.fecal.genus2

#Merge phyloseq objects
amaze.microbiome <- merge_phyloseq(amaze.fecal.genus2, amaze.duod.genus2)
amaze.microbiome
otu_table(amaze.microbiome)[1:5, 1:5]
```

add phylogenetic tree
```{r}
###Create phylogenetic tree
library(seqinr)
library(ape)
sequence <- refseq(amaze.microbiome)  
## Perform multiple sequence alignment
my_alignment <- msa::msa(sequence)

## Compute distance matrix
my_alignment_sequence <- msa::msaConvert(my_alignment, type="seqinr::alignment")

distance_alignment <- dist.alignment(my_alignment_sequence)

## compute phylogenetic tree using neighbor joining
Tree <- bionj(distance_alignment)

###Add tree to phyloseq object
phy_tree(amaze.microbiome) <- Tree
amaze.microbiome
```

Clean phyloseq object
```{r}
#Remove unwanted samples
amaze.microbiome <- subset_samples(amaze.microbiome, InterventionArm != "")
amaze.microbiome

#consistent naming
levels(as.factor(amaze.microbiome@sam_data[["InterventionArm"]]))
amaze.microbiome@sam_data[["InterventionArm"]] <- ifelse(amaze.microbiome@sam_data[["InterventionArm"]]== "Micronutreint","Micronutrient", amaze.microbiome@sam_data[["InterventionArm"]])

#exclude point with missing data
ps <- subset_samples(amaze.microbiome, InterventionArm != "")

#remove taxa with no abundance data
ps_filtered <- prune_taxa(taxa_sums(ps) > 0, ps)
ps_filtered

#Make tree rooted: this is needed to convert to microbiomeStat object
ape::is.rooted(phy_tree(ps_filtered)) 
phy_tree(ps_filtered) <- ape::root(phy_tree(ps_filtered), sample(taxa_names(ps_filtered), 1), resolve.root = TRUE)
ape::is.rooted(phy_tree(ps_filtered)) 
ps_filtered
```

Save combined data
```{r}
save(ps_filtered, file = "Inputs/RData/AMAZE_ps_filtered.RData")
```

create input for picrust2
```{r}
#1. merge dataset at ASV level
load("Inputs/RData/AMAZE_duodenum.RData") #ps0_duodenum
load("Inputs/RData/AMAZE_fecal.RData") #ps0_fecal
library(biomformat);packageVersion("biomformat")
library(phyloseq)
library(tidyverse)
#2. export count table with ASV id as the first column
#a. duodenum
tax_duod <- as(tax_table(ps0_duodenum),"matrix")%>%
  as.data.frame() %>% rownames_to_column(var = "rownam")
otu_duod <-as(otu_table(ps0_duodenum),"matrix") %>% # 't' to transform if taxa_are_rows=FALSE
  as.data.frame() %>% rownames_to_column(var = "rownam") 
otu_duod <- right_join(tax_duod %>% select(rownam, ASVSeq), otu_duod) %>% 
  column_to_rownames(var = "ASVSeq") %>%  select(-rownam)
tax_duod <- tax_duod %>% column_to_rownames(var = "ASVSeq") %>% 
  select(Kingdom, Phylum, Class, Order, Family, Genus, Species)  
#b. stool
tax_stool <- as(tax_table(ps0_fecal),"matrix")%>%
  as.data.frame()%>% rownames_to_column(var = "rownam")
otu_stool <- t(as(otu_table(ps0_fecal),"matrix")) %>% # 't' to transform if taxa_are_rows=FALSE
  as.data.frame()%>% rownames_to_column(var = "rownam") 
otu_stool <- right_join(tax_stool %>% select(rownam, ASVSeq), otu_stool) %>%
  column_to_rownames(var = "ASVSeq") %>% select(-rownam)
tax_stool <- tax_stool %>% column_to_rownames(var = "ASVSeq") %>% 
  select(Kingdom, Phylum, Class, Order, Family, Genus, Species)  

amaze_otu <- merge(otu_duod, otu_stool, by = "row.names", all = TRUE)
#c. Merge
amaze_tax <- full_join(tax_duod %>% rownames_to_column(var = "ASVSeq"), 
                   tax_stool  %>% rownames_to_column(var = "ASVSeq"), 
                   by = c("ASVSeq")) %>% #%>%  column_to_rownames(var = "ASVSeq")
  select(ASVSeq, Kingdom=Kingdom.x, Phylum=Phylum.x , Class=Class.x, Order=Order.x, Family=Family.x, Genus=Genus.x, Species=Species.x)  %>% 
  mutate(ASV = paste("ASV",1:n(), sep = "_"))
#d. rename otu table to ASVids
amaze_otu2 <- full_join(amaze_otu %>% 
                          rename('Row.names'='ASVSeq'), 
                        amaze_tax %>% 
                          select(ASV, ASVSeq), 
                        by = "ASVSeq")%>%
  select(-ASVSeq) %>% 
  column_to_rownames(var = "ASV") %>% 
  replace(is.na(.), 0) %>% 
  as.matrix() 
otu_biom<-make_biom(data=amaze_otu2)
write_biom(otu_biom,"Inputs/Picrust2/AMAZE_biom.biom")

#2. Put sequences for each ASV in fasta file
library(scifer)
df_to_fasta(
    sequence_name = amaze_tax$ASV,
    sequence_strings = amaze_tax$ASVSeq,
    file_name = "AMAZE_sequences.fna",
    output_dir = "Inputs/Picrust2",
    save_fasta = TRUE
)

#save tax atble
write_csv(amaze_tax, "Inputs/Picrust2/AMAZE_tax.csv")
```


