---
title: "AMAZE - Baseline Characteristics (stool vs Duoenal Microbiome)"
Author: Monica N Mweetwa
Date: 17/04/2024
---
1. Set working directory
```{r setup}
    knitr::opts_knit$set(root.dir = normalizePath("C:/Users/Monica/Dropbox/AMAZE")) 
    #knitr::opts_knit$set(root.dir = normalizePath("~/Library/CloudStorage/Dropbox/AMAZE")) 
```

2. Load necessary packages 
```{r include=FALSE}
pacman::p_load(phyloseq, 
               Maaslin2,
               ggplot2, 
               ggpubr,
               tidyverse,
               microViz,
               MicrobiomeStat)
source("Code/3. SupplementationAnalysis/OrdPlot_Functions.R")
```

3. Load Data
```{r}
load("Inputs/RData/AMAZE_ps_filtered.RData")
ps_filtered
amaze_dup2 <- subset_samples(ps_filtered, Timepoint == "Pre-Intervention")
```

4. Analysis

A. Beta Diversity
```{r}
set.seed(1234)
#permanova test on bray curtis distances
library(vegan)
dist <- phyloseq::distance(amaze_dup2, method="bray")
metadata <- data.frame(sample_data(amaze_dup2))
adonis2(dist ~ Sample.Type, data = metadata, na.action = na.omit) #pval = 0.001
  
beta.plot1 <- amaze_dup2 %>%
  tax_transform("clr", rank = "Genus") %>%
  ord_calc(method = "PCA") %>%
  ord_plot(color = "Sample.Type", plot_taxa = 1:20, size = 5, auto_caption = NA, scaling = "none", expand = TRUE,
           tax_lab_style = tax_lab_style(type = "text", size = 4, check_overlap = TRUE)) +
  scale_colour_brewer(palette = "Dark2") +
  labs(color = "Sample Type", shape = "Timepoint") +
  theme_bw()
#Add PERMANOVA to beta.plot2
beta.plot22 <- beta.plot1 + 
  annotate("text", x = 0.01, y = 0.2, 
  label = paste("PERMANOVA Test: p value = 0.001"),
  size = 5) +
  scale_colour_brewer(palette = "Dark2") +
  scale_fill_brewer(palette = "Dark2") 
ggsave("Outputs/2. Baseline/SampleType/BetaPlot.SampleType.png", plot = beta.plot22, dpi = 600, device = "png", width = 30, height = 15, units = "cm")
#save for combined plots
save(beta.plot22, file = "Inputs/RData/ForCombinedPlots/BetaPlot_SampleType.RData")
```

B. Differential abundance between duodenal and stool samples
```{r}
mas_P <- Maaslin2(
  input_data = data.frame(otu_table(amaze_dup2)),
  input_metadata = data.frame(sample_data(amaze_dup2)),
  output = "Outputs/2. Baseline/SampleType/Diff_Abund",
  min_abundance = 0.001,
  min_prevalence = 0.05,
  normalization = "TMM",
  transform = "NONE",
  analysis_method = "NEGBIN",
  max_significance = 0.1,
  fixed_effects = c("Sample.Type"),
  random_effects = "PID",
  correction = "BY",
  standardize = FALSE,
  cores = 1,
  plot_heatmap = TRUE,
  plot_scatter = TRUE,
  heatmap_first_n = 50)
pp.stool <- mas_P[["results"]]
notsig <- pp.stool %>%
  mutate(sam.type = ifelse(.$coef > 0,  "Stool", "Duodenal Aspirate"),
         sam.type = (ifelse(.$qval > 0.05 , "Common", sam.type)),
         coef.exp = exp(coef),
         logfold2 = log2(coef.exp),
         logfold2.2 = ifelse(logfold2 < 0, logfold2*-1, logfold2)) 

write_csv(notsig, "Outputs/2. Baseline/SampleType/lda_sam.csv")

SigBact <- pp.stool %>%
  mutate(sam.type = ifelse(.$coef > 0,  "Stool", "Duodenal Aspirate"),
         sam.type = (ifelse(.$qval > 0.05 , "Common", sam.type)),
         coef.exp = exp(coef),
         logfold2 = log2(coef.exp),
         logfold2.2 = ifelse(logfold2 < 0, logfold2*-1, logfold2)) %>%
  filter(qval < 0.05)
write_csv(SigBact, "Outputs/2. Baseline/SampleType/lda_sam_unique.csv")

lda.sam <- ggpubr::ggbarplot(SigBact, x = "feature", y = "logfold2.2",
          fill = "sam.type",           # change fill color by mpg_level
          color = "white",            # Set bar border colors to white
          palette = "Dark2",            # jco journal color palett. see ?ggpar
          sort.val = "asc",           # Sort the value in ascending order
          sort.by.groups = T,     # Don't sort inside each group
          x.text.angle = 90,          # Rotate vertically x axis texts
          ylab = "Log2Fold Change",
          xlab = "Taxon",
          legend.title = "Sample Type",
          rotate = TRUE,
          ggtheme = theme_classic()
          ) 
ggsave("Outputs/2. Baseline/SampleType/lda.sam.png", plot = lda.sam, dpi = 600, device = "png", width = 18, height = 15, units = "cm")
save(lda.sam, file = "Inputs/RData/ForCombinedPlots/Diff_AbundBaseline_SampleType.RData")
```

C. Pathway differences 
This analysis take output from picrust2 as input that was run in a conda environment (picrust-src) using the follwoing command:
picrust2_pipeline.py -s ~/Dropbox/AMAZE/Inputs/Picrust2/AMAZE_sequences.fna \
-i ~/Dropbox/AMAZE/Inputs/Picrust2/AMAZE_biom.biom \
-o ~/Dropbox/AMAZE/Outputs/'4. Picrust2'/AMAZE_out_picrust -p 1

This produced metagenome and pathway predictions based on the abundance of 16 taxa in our dataset. This can be analysed similar to taxonomic abundance analysis using ggpicrust2

```{r}
library(MicrobiomeStat)
library(MicrobiomeProfiler)
library(ggpicrust2)
library(readr)
library(Maaslin2)
#import data
#pathways abundance
pathway_dat <- read.delim("Outputs/4. Picrust2/AMAZE_out_picrust/pathways_out/path_abun_unstrat.tsv")%>%
  column_to_rownames(var = "pathway")
names(pathway_dat) <- substring(names(pathway_dat),2)

#metadata
metadata <- read_csv("Inputs/Metadata/metadata_paired_withELISA.csv") %>%
  column_to_rownames(var = "SampleID2") %>%
  filter(Timepoint == "Pre-Intervention")

#Get pathway descriptions and create feature.ann (pathway taxonomy)
path_df <- pathway_annotation(
  file = "Outputs/4. Picrust2/AMAZE_out_picrust/pathways_out/path_abun_unstrat.tsv",
  pathway = "MetaCyc",
  daa_results_df = path_df,
  ko_to_kegg = FALSE
)
#add pathway description to file
feature_ann <- path_df %>% 
  select(pathway, description) %>% 
  mutate(pathID = pathway) %>%
  column_to_rownames(var = "pathway")

#create MicrobiomeStat object
ps.obj <- list(
  feature.tab = as.matrix(pathway_dat),
  meta.dat = metadata,
  feature.ann = as.matrix(feature_ann)
)

# Generate taxa test
test.list <- generate_taxa_test_single(
    data.obj = ps.obj,
    time.var = "Timepoint",
    t.level = "Pre-Intervention",
    group.var = "Sample.Type",
    #adj.vars = "sex",
    feature.dat.type = "count",
    feature.level = c("pathID"),
    prev.filter = 0.1,
    abund.filter = 0.0001
)

# Generate volcano plots
volcano_plots <- generate_taxa_volcano_single(
    data.obj = ps.obj,
    group.var = "Sample.Type",
    test.list = test.list,
    feature.sig.level = 0.05,
    feature.mt.method = "none"
)

v1 <- volcano_plots[["pathID"]][["Stool vs Duodenal Aspirate (Reference)"]] +
  theme(legend.position = "right")
ggsave(plot = v1, filename = "Outputs/2. Baseline/SampleType/pathwayplot2.png", dpi = 600)
```

################ do not run code below ####################################


C. Associations between stool and duodenal microbiome
```{r}
load("Inputs/AMAZE_ps_filtered_pairs.RData")
amaze_dup2 <- subset_samples(amaze_dup, Timepoint == "Pre-Intervention")

tax_t <- as.data.frame(amaze_dup2@otu_table) 
met_t <- as.data.frame(amaze_dup2@sam_data)


tax_2 <- tax_t %>%
  rownames_to_column(var = "Sample")
q <- ps_melt(amaze_dup2)
qs <- q%>%
  select(Sample, PID, InterventionArm, Timepoint, hiv_status, sex_c, Sample.Type) %>%
  #filter( %>%
  unique()

for_plot <- left_join(qs, tax_2)
for_plot2 <- for_plot %>%
  select(-Sample) %>%
  pivot_wider(names_from = Sample.Type, 
              values_from = -c(PID, InterventionArm, Timepoint, hiv_status, sex_c, Sample.Type),
              names_glue = "{.value}_{Sample.Type}")
for_plot2[is.na(for_plot2)] <- 0

library(corrplot)
for_corr <- for_plot2 %>%
  select(-c(PID, InterventionArm, Timepoint, hiv_status, sex_c)) %>%
  select(where(~ sum(.) != 0))


#rename taxa
num_rows <- 394
seq_let <- rep (letters, length.out = num_rows)

forcorrr_t <- t(for_corr)
forcorr2 <- forcorrr_t %>%
  as.data.frame()%>%
  rownames_to_column(var= "tax") %>%
  mutate(let_col = paste0("tax_",1:n()))
write.csv(forcorr2 %>%
            select(tax, let_col), "Outputs/2. Baseline/SampleType/microbe_names.csv")

forcorr3 <- forcorr2 %>%
  select(-tax)  %>%
  column_to_rownames(var = "let_col")
forcorr2_t <- t(forcorr3)  


#Calculate correlation coefficient
library(psych)
library(tidyverse)
adj.dat <- corr.test(forcorr2_t, method = "spearman", adjust = "holm")
write.csv(adj.dat$ci, "Outputs/2. Baseline/SampleType/microbe_cor_pval.csv")
write.csv(adj.dat$ci2, "Outputs/2. Baseline/SampleType/microbe_cor_pval_adjust.csv")
write.csv(adj.dat$r, "Outputs/2. Baseline/SampleType/microbe_cor_rvalues.csv")

dat3 <- adj.dat$ci %>%
  rownames_to_column(var = "tax0") %>%
  separate(tax0, into = c("tax1", "tax2"), sep = "-") %>%
  separate(tax1, into = c("tx01", "tx1"), sep = "_") %>%
  separate(tax2, into = c("tx02", "tx2"), sep = "_") %>%
  select(-tx01, -tx02) %>%
  filter(p < 0.05) %>%
  filter(p != 0)

dd2 <- forcorr2 %>%
  select(tax, let_col) %>%
  separate(let_col, into = c("tx01", "tx1"), sep = "_") %>%
  select(-tx01)

data_fin <- merge(dat3, dd2, by = "tx1", all = T)

dd3 <- forcorr2 %>%
  select(tax, let_col) %>%
  separate(let_col, into = c("tx01", "tx2"), sep = "_") %>%
  select(-tx01) %>%
  rename(tax2 = tax)
data_fin2 <- merge(data_fin, dd3, by = "tx2", all = T)

term1 <- "Stool"
term2 <- "Duodenal Aspirate"
data_fin2 <- data_fin2%>%
  filter(!is.na(tx2)) %>%
  filter((str_detect(tax, term1) & str_detect(tax2, term2)) |
           (str_detect(tax, term2) & str_detect(tax2, term1)))

write.csv(data_fin2, "Outputs/2. Baseline/SampleType/microbe_cor_sig_duod_stool.csv")
```


Alternative pathwya analysis
```{r}
library(MicrobiomeProfiler)
library(ggpicrust2)
library(readr)
library(Maaslin2)
#import data
#pathways abundance
pathway_dat <- read.delim("Outputs/4. Picrust2/AMAZE_out_picrust/pathways_out/path_abun_unstrat.tsv")%>%
  column_to_rownames(var = "pathway")
names(pathway_dat) <- substring(names(pathway_dat),2)

#metadata
metadata <- read_csv("Inputs/Metadata/metadata_paired_withELISA.csv") %>%
  column_to_rownames(var = "SampleID2") %>%
  filter(Timepoint == "Pre-Intervention")

#Get pathway descriptions and create feature.ann (pathway taxonomy)
library(ggpicrust2)
path_df <- pathway_annotation(
  file = "Outputs/4. Picrust2/AMAZE_out_picrust/pathways_out/path_abun_unstrat.tsv",
  pathway = "MetaCyc",
  daa_results_df = path_df,
  ko_to_kegg = FALSE
)
#add pathway description to file
feature_ann <- path_df %>% 
  select(pathway, description) %>% 
  mutate(pathID = pathway) %>%
  column_to_rownames(var = "pathway")

pathway_dat2 <- pathway_dat %>% select(contains("A1"), contains("S1"))
mas_P <- Maaslin2(
  input_data = pathway_dat2,
  input_metadata = metadata,
  output = "Outputs/2. Baseline/SampleType/Picrust2",
  min_abundance = 0.001,
  min_prevalence = 0.05,
  normalization = "TMM",
  transform = "NONE",
  analysis_method = "NEGBIN",
  max_significance = 0.1,
  fixed_effects = c("Sample.Type"),
  correction = "BY",
  standardize = FALSE,
  cores = 1,
  plot_heatmap = TRUE,
  plot_scatter = TRUE,
  heatmap_first_n = 50)
res <- mas_P[["results"]]
notsig <- res %>%
  mutate(sam.type = ifelse(.$coef > 0,  "Stool", "Duodenal Aspirate"),
         sam.type = (ifelse(.$qval > 0.05 , "Common", sam.type)),
         coef.exp = exp(coef),
         logfold2 = log2(coef.exp),
         logfold2.2 = ifelse(logfold2 < 0, logfold2*-1, logfold2),
         log10_pvalue = log10(qval)) 
#replace the dots with high-phen
notsig2 <- notsig %>%
  mutate(feature = str_replace_all(feature, "\\.", "-"))

write_csv(notsig2, "Outputs/2. Baseline/SampleType/lda_pathway.csv")

#pathway plot#
library(ggplot2)
ggplot(notsig2 %>% filter(qval != 0), aes(x = logfold2, y = -log10_pvalue, size = N.not.zero)) +
  geom_point(aes(color = sam.type)) +
  scale_color_manual(values = c("Common" = "gray", "Duodenal Aspirate" = "red", "Stool" = "red") ) +
  theme(legend.position = "none") +
  theme_classic()

```



Duodenal and Stool microbial correlations
```{r}
load("Inputs/RData/AMAZE_ps_filtered.RData")

library(ggplot2)
library(phyloseq)
library(tidyverse)
library(microbiome)
library(microViz)
amaze_dup2 <- subset_samples(ps_filtered, Timepoint == "Pre-Intervention")
# Define the threshold: 20% of the total number of samples
threshold <- 0.2 * nsamples(amaze_dup2)
# Filter taxa based on presence in at least 20% of samples
physeq_filtered <- prune_taxa(taxa_sums(amaze_dup2) > 0 & rowSums(otu_table(amaze_dup2) > 0) >= threshold, amaze_dup2)
# physeq_filtered now contains only the taxa present in at least 20% of the samples

tax_t <- as.data.frame(physeq_filtered@otu_table) 
met_t <- as.data.frame(physeq_filtered@sam_data)


tax_2 <- tax_t %>%
  rownames_to_column(var = "Sample")
q <- ps_melt(amaze_dup2)
qs <- q%>%
  select(Sample, PID, InterventionArm, Timepoint, hiv_status, sex_c, Sample.Type) %>%
  #filter( %>%
  unique()

for_plot <- left_join(qs, tax_2)
for_plot2 <- for_plot %>%
  select(-Sample) %>%
  pivot_wider(names_from = Sample.Type, 
              values_from = -c(PID, InterventionArm, Timepoint, hiv_status, sex_c, Sample.Type),
              names_glue = "{.value}_{Sample.Type}")
for_plot2[is.na(for_plot2)] <- 0

library(corrplot)
for_corr <- for_plot2 %>%
  select(-c(PID, InterventionArm, Timepoint, hiv_status, sex_c)) %>%
  select(where(~ sum(.) != 0))


#rename taxa
num_rows <- 394
seq_let <- rep (letters, length.out = num_rows)

forcorrr_t <- t(for_corr)
forcorr2 <- forcorrr_t %>%
  as.data.frame()%>%
  rownames_to_column(var= "tax") %>%
  mutate(let_col = paste0("tax_",1:n()))
write.csv(forcorr2 %>%
            select(tax, let_col), "Outputs/2. Baseline/SpearmanCorr/microbe_names.csv")

forcorr3 <- forcorr2 %>%
  select(-tax)  %>%
  column_to_rownames(var = "let_col")
forcorr2_t <- t(forcorr3)  

#Calculate correlation coefficient
library(psych)
library(tidyverse)
adj.dat <- corr.test(forcorr2_t, method = "spearman", adjust = "BY")
write.csv(adj.dat$ci, "Outputs/2. Baseline/SpearmanCorr/microbe_cor_pval.csv")
write.csv(adj.dat$ci2, "Outputs/2. Baseline/SpearmanCorr/microbe_cor_pval_adjust.csv")
write.csv(adj.dat$r, "Outputs/2. Baseline/SpearmanCorr/microbe_cor_rvalues.csv")

dat3 <- adj.dat$ci %>%
  rownames_to_column(var = "tax0") %>%
  separate(tax0, into = c("tax1", "tax2"), sep = "-") %>%
  separate(tax1, into = c("tx01", "tx1"), sep = "_") %>%
  separate(tax2, into = c("tx02", "tx2"), sep = "_") %>%
  select(-tx01, -tx02) %>%
  filter(p < 0.05) %>%
  filter(p != 0)

dd2 <- forcorr2 %>%
  select(tax, let_col) %>%
  separate(let_col, into = c("tx01", "tx1"), sep = "_") %>%
  select(-tx01)

data_fin <- merge(dat3, dd2, by = "tx1", all = T)

dd3 <- forcorr2 %>%
  select(tax, let_col) %>%
  separate(let_col, into = c("tx01", "tx2"), sep = "_") %>%
  select(-tx01) %>%
  rename(tax2 = tax)
data_fin2 <- merge(data_fin, dd3, by = "tx2", all = T)

term1 <- "Stool"
term2 <- "Duodenal Aspirate"
data_fin2 <- data_fin2%>%
  filter(!is.na(tx2)) %>%
  filter((str_detect(tax, term1) & str_detect(tax2, term2)) |
           (str_detect(tax, term2) & str_detect(tax2, term1)))

write.csv(data_fin2, "Outputs/2. Baseline/microbe_cor_sig_duod_stool.csv")

#Correlation plot
#filter correlation matrix for correlations between duodenal and stool microbes only
corrs <- readxl::read_excel("Outputs/2. Baseline/SpearmanCorr/microbe_corrplot.xlsx")
#mat <- adj.dat$r
#filtered_mat <- mat[rownames(mat) %in% corrs$tx, ]
#filtered_mat2 <- mat[, colnames(filtered_mat) %in% corrs$tx2]

# Plot
library(ggplot2)
cr_plot <- ggplot(data = corrs, aes(x=fct_reorder(Duod,r), y=fct_reorder(Stool,r), fill=r)) + 
  geom_tile(color = "white")+
 scale_fill_gradient2(low = "blue", high = "red", mid = "white", 
   midpoint = 0, limit = c(-1,1), space = "Lab", 
   name="Spearson\nCorrelation") +
  theme_minimal() + 
  labs(y = "Stool Genera", x = "Duodenal Genera") +
 theme(axis.text.x = element_text(angle = 45, vjust = 1, 
    size = 12, hjust = 1)) +
 coord_fixed()

ggsave("Outputs/2. Baseline/SuppFig6.png", plot = cr_plot, dpi = 300, device = "png", width = 55, height = 75, units = "cm")
```


