---
title: "Stool AMAZE Data Cleaning - Monica N Mweetwa"
---
Set working directory
```{r, setup}
knitr::opts_knit$set(root.dir = 'C:/Users/Monica/Dropbox/AMAZE/')
#knitr::opts_knit$set(root.dir = '~/Dropbox/AMAZE')
```

Load packages
```{r include=FALSE}
library("phyloseq")
library("vegan")
library("tidyverse")
library(seqinr)
library(ape)
library(msa) 
library(stringr)
library(upstartr)
library(microbiome)
```

Load Fecal Data
```{r include=FALSE}
load("Inputs/Fecal/2023-08-16_Amaze_fecal_16S_rDNA_data.RData")
TAXtable <- read_csv("Inputs/Fecal/2023-8-16_AMAZE_TAXtable.csv")
```

There already is a phyloseq object called ps
Add sequence information to phyloseq object using the taxonomy tables
```{r}
TAXtable.1 <- TAXtable%>%
  select(...1, ASVSeq)
TAXtable.1$fasta_name <- paste0(">",TAXtable.1$...1)
TAXtable.1 <- TAXtable.1%>%
  select(fasta_name, ASVSeq)%>%
  rename(fasta_name = "seq.name",
         ASVSeq = "seq.text")
library(phylotools)
dat2fasta(TAXtable.1, outfile = "Inputs/Fecal/AMAZE_stool_sequences.fasta")

```

Create phylogenetic tree
```{r}
sequence_fecal <- readAAStringSet("Inputs/Fecal/AMAZE_stool_sequences.fasta")   
## Perform multiple sequence alignment
my_alignment_f <- msa(sequence_fecal)

## Compute distance matrix
my_alignment_f_sequence <- msaConvert(my_alignment_f, type="seqinr::alignment")

distance_alignment_f <- dist.alignment(my_alignment_f_sequence)

## compute phylogenetic tree using neighbor joining
Tree <- bionj(distance_alignment_f)
```

Add tree to phyloseq object
```{r}
#Rename tip.labels of tree to match the names of phyloseq taxa
Tree2 <- Tree
Tree2$tip.label <- str_sub(Tree2$tip.label, start = 2)

write.tree(Tree2, "Inputs/Fecal/amaze_tree_fecal.tre")

phy_tree(ps) <- Tree2
ps
```

Add sequence data
```{r}
sequence_fecal2 <- sequence_fecal
sequence_fecal2@ranges@NAMES <- str_sub(sequence_fecal2@ranges@NAMES, start = 2)

ps1 <- merge_phyloseq(ps, sequence_fecal2)
ps1
```

Add metadata to phyloseq object
```{r include=FALSE}
#1. Load additional data
library(readr)
sample_metadata <- read.csv("Inputs/Fecal/metdat_stool.csv")

##change order so that bar plots show pre-intervention before post-interventions
levels(as.factor(sample_metadata$Timepoint))
sample_metadata$Timepoint <- factor(sample_metadata$Timepoint, levels = c("Pre-Intervention", "Post-Intervention", ""))
levels(as.factor(sample_metadata$Timepoint))

#Add metadata 
sample_id <- sample_metadata %>% 
  tibble::column_to_rownames("SampleID") 
sample_data(ps1) <- sample_id
ps0_fecal <- ps1
ps0_fecal
```

Clean phyloseq object
```{r}
ps0 <- subset_samples(ps0_fecal, InterventionArm != "") #Remove samples that are not paired 
ps0 = subset_taxa(ps0, Kingdom == "Bacteria") # remove those with no taxonomic assignment
ps0 = subset_taxa(ps0, Family != "Mitochondria") #remove those assigned as Mitochondria Family
ps0 = subset_taxa(ps0, Class != "Chloroplast") #remove those assigned as Chloroplast Class 
ps0
```

Save RData file
```{r}
save(ps0_fecal, file = "Inputs/RData/AMAZE_fecal.RData")
```

Agglomerate at Genus level
```{r}
ps0.genus.fecal <- tax_glom(ps0_fecal, taxrank = "Genus", NArm = TRUE)
#Rename taxa with genus names
genus.names <- make.names(tax_table(ps0.genus.fecal)[,'Genus'], unique = TRUE)
taxa_names(ps0.genus.fecal) <- genus.names
ps0.genus.fecal 
```
Save RData file
```{r}
save(ps0.genus.fecal, file = "Inputs/RData/AMAZE_fecal_genus.RData")
```
